bindings/go/dag/dag.go

```diff
@@ -0,0 +1,328 @@
+	}
+}
+
+// AddVertex adds a new node to the graph.
+func (d *DirectedAcyclicGraph[T]) AddVertex(id T, attributes ...map[string]any) error {
+	if _, exists := d.Vertices[id]; exists {
+		return fmt.Errorf("node %v already exists", id)
+	}
+	d.Vertices[id] = &Vertex[T]{
+		ID:         id,
+		Attributes: make(map[string]any),
+		Edges:      make(map[T]map[string]any),
+	}
+
+	for _, attributes := range attributes {
+		maps.Copy(d.Vertices[id].Attributes, attributes)
+	}
+
+	d.OutDegree[id] = 0
+	d.InDegree[id] = 0
+	return nil
+}
+
+// DeleteVertex removes a node from the graph.
+func (d *DirectedAcyclicGraph[T]) DeleteVertex(id T) error {
+	if _, exists := d.Vertices[id]; !exists {
+		return fmt.Errorf("node %v does not exist", id)
+	}
+
+	// Remove all edges to this node
+	for _, node := range d.Vertices {
+		if _, exists := node.Edges[id]; exists {
+			// Decrement the in-degree of the node
+			d.InDegree[node.ID]--
+			// Remove the edge from the node
+			delete(node.Edges, id)
+		}
+	}
+
+	delete(d.Vertices, id)
+	delete(d.OutDegree, id)
+	delete(d.InDegree, id)
+
+	return nil
+}
+
+type CycleError struct {
+	Cycle []string
+}
+
+func (e *CycleError) Error() string {
+	return fmt.Sprintf("The current graph would create a cycle: %s", formatCycle(e.Cycle))
+}
+
+func formatCycle(cycle []string) string {
+	return strings.Join(cycle, " -> ")
+}
+
+// AddEdge adds a directed edge from one node to another.
+func (d *DirectedAcyclicGraph[T]) AddEdge(from, to T, attributes ...map[string]any) error {
+	fromNode, fromExists := d.Vertices[from]
+	_, toExists := d.Vertices[to]
+	if !fromExists {
+		return fmt.Errorf("node %v does not exist", from)
+	}
+	if !toExists {
+		return fmt.Errorf("node %v does not exist", to)
+	}
+	if from == to {
+		return ErrSelfReference
+	}
+
+	_, exists := fromNode.Edges[to]
+
+	if !exists {
+		// Only initialize the map if the edge was added
+		fromNode.Edges[to] = map[string]any{}
+		// Only increment the out-degree and in-degree if the edge was added
+		d.OutDegree[from]++
+		d.InDegree[to]++
+
+		// Check if the graph is still a DAG
+		hasCycle, cycle := d.HasCycle()
+		if hasCycle {
+			// Ehmmm, we have a cycle, let's remove the edge we just added
+			delete(fromNode.Edges, to)
+			d.OutDegree[from]--
+			d.InDegree[to]--
+
+			return fmt.Errorf("adding an edge from %v to %v would create a cycle: %w", fmt.Sprintf("%v", from), fmt.Sprintf("%v", to), &CycleError{
+				Cycle: cycle,
+			})
+		}
+	}
+
+	for _, attributes := range attributes {
+		maps.Copy(fromNode.Edges[to], attributes)
+	}
+
+	return nil
+}
+
+func (d *DirectedAcyclicGraph[T]) Roots() []T {
+	var roots []T
+	for node, degree := range d.InDegree {
+		if degree == 0 {
+			roots = append(roots, node)
+		}
+	}
+	return roots
+}
+
+func (d *DirectedAcyclicGraph[T]) TopologicalSort() ([]T, error) {
+	if cyclic, nodes := d.HasCycle(); cyclic {
+		return nil, &CycleError{
+			Cycle: nodes,
+		}
+	}
+
+	visited := make(map[T]bool)
```

Comment:
I'd prefer using `struct{}{}`. But not a hard stopper. :)

Reply:
map to bool in this case is just a nicer interface, and I dont think we optimize for that much (yet?)

Reply:
Yeah, fine. :D